src/rhsmlib/facts/cloud_facts.py

```diff
@@ -51,17 +53,27 @@ def __init__(self, arch=None, prefix=None, testing=None, collected_hw_info=None)
 
     # https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/identify_ec2_instances.html
     def get_aws_instance_id(self):
+        """
+        Try to get instance ID of machine running on AWS
+        :return: dictionary containing {"aws_instance_id": some_instance_ID}, when the machine is running on
+            AWS cloud; otherwise returns empty dictionary {}
+        """
         values = {}
-        if not self._collected_hw_info['dmi.bios.version'].find('amazon') >= 0:
-            return ""
+
+        # Is the machine running on VM and is it AWS?
+        aws_cloud_detector = AWSCloudDetector(self._collected_hw_info)
+        if aws_cloud_detector.is_running_on_cloud() is False:
+            return {}
+
+        # Try to read instance ID from cache first
```

Comment:
I realize this should probably have been a question on the PR that originally put the facts_cache into use for the instance ID check but, can the instance ID change? Like say for example a system is started, we cache the instance ID, then that system is cloned. In that case would the instance ID in actuality match the one we report?

Reply:
Good question. I really don't know. I will ask RHUI team. They could know answer on the question if machine ID could change over time.

Reply:
We can circle back to this, but I think it is worth us removing the caching of the instance id, it's another point of failure in migration cases or in case where a running image / filesystem is cloned (unless we are clearing the cache on unregistration).

Reply:
Sure, I will address this issue in PR related to collecting metadata.